libs/Microsoft.MixedReality.WebRTC/AudioTransceiver.cs

```diff
@@ -0,0 +1,146 @@
+        /// 
+        /// This detaches the previous local audio track if any, and attaches the new one instead.
+        /// Note that the transceiver will only send some audio data to the remote peer if its
+        /// negotiated direction includes sending some data and it has an attached local track to
+        /// produce this data.
+        /// 
+        /// This change is transparent to the session, and does not trigger any renegotiation.
+        /// </summary>
+        /// <param name="track">The new local audio track attached to the transceiver, and used to
+        /// produce audio data to send to the remote peer if the transceiver is sending.
+        /// Passing <c>null</c> is allowed, and will detach the current track if any.</param>
+        public void SetLocalTrack(LocalAudioTrack track)
+        {
+            if (track == _localTrack)
+            {
+                return;
+            }
+
+            if (track != null)
+            {
+                if ((track.PeerConnection != null) && (track.PeerConnection != PeerConnection))
+                {
+                    throw new InvalidOperationException($"Cannot set track {track} of peer connection {track.PeerConnection} on audio transceiver {this} of different peer connection {PeerConnection}.");
+                }
+                var res = TransceiverInterop.Transceiver_SetLocalAudioTrack(_nativeHandle, track._nativeHandle);
+                Utils.ThrowOnErrorCode(res);
+            }
+            else
+            {
+                // Note: Cannot pass null for SafeHandle parameter value (ArgumentNullException)
+                var res = TransceiverInterop.Transceiver_SetLocalAudioTrack(_nativeHandle, new LocalAudioTrackHandle());
+                Utils.ThrowOnErrorCode(res);
+            }
+
+            // Capture peer connection; it gets reset during track manipulation below
+            var peerConnection = PeerConnection;
+
+            // Remove old track
+            if (_localTrack != null)
+            {
+                _localTrack.OnTrackRemoved(peerConnection);
+            }
+            Debug.Assert(_localTrack == null);
+
+            // Add new track
+            if (track != null)
+            {
+                track.OnTrackAdded(peerConnection, this);
+                Debug.Assert(track == _localTrack);
+                Debug.Assert(_localTrack.PeerConnection == PeerConnection);
+                Debug.Assert(_localTrack.Transceiver == this);
+                Debug.Assert(_localTrack.Transceiver.LocalTrack == _localTrack);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnLocalTrackMuteChanged(bool muted)
+        {
+            LocalTrack?.OnMute(muted);
+        }
+
+        /// <inheritdoc/>
+        protected override void OnRemoteTrackMuteChanged(bool muted)
+        {
+            RemoteTrack?.OnMute(muted);
+        }
+
+        internal void OnLocalTrackAdded(LocalAudioTrack track)
```

Comment:
This should be called "SetLocalTrackInternal" or something similar, the `On` is conventionally used for callbacks and this is called normally by other methods.

In general all these methods called "On[Local|Remote]TrackAdded" in various classes sound mis-named, the name doesn't say what they actually do and it's impossible to figure out which calls which.

Reply:
Fixed.